src/Elmish.WPF/BindingData.fs

```diff
@@ -134,30 +135,115 @@ and internal AlterMsgStreamData<'model, 'msg, 'bindingModel, 'bindingMsg, 'dispa
+          Set = fIn >> d.Set
+        }
+      | OneWaySeqData d -> OneWaySeqData {
+          Get = d.Get
+          CreateCollection = d.CreateCollection >> CollectionTarget.mapVm fOut
+          GetId = d.GetId
+          ItemEquals = d.ItemEquals
+        }
+      | TwoWayData d -> TwoWayData {
+          Get = d.Get >> fOut
+          Set = fIn >> d.Set
+        }
+      | CmdData d -> CmdData {
+          Exec = d.Exec
+          CanExec = d.CanExec
+          AutoRequery = d.AutoRequery
+        }
+      | SubModelData d -> SubModelData {
+          GetModel = d.GetModel
+          CreateViewModel = d.CreateViewModel >> fOut
+          UpdateViewModel = (fun (vm,m) -> d.UpdateViewModel (fIn vm, m))
+          ToMsg = d.ToMsg
+        }
+      | SubModelWinData d -> SubModelWinData {
+          GetState = d.GetState
+          CreateViewModel = d.CreateViewModel >> fOut
+          UpdateViewModel = (fun (vm,m) -> d.UpdateViewModel (fIn vm, m))
+          ToMsg = d.ToMsg
+          GetWindow = d.GetWindow
+          IsModal = d.IsModal
+          OnCloseRequested = d.OnCloseRequested
+        }
+      | SubModelSeqUnkeyedData d -> SubModelSeqUnkeyedData {
+          GetModels = d.GetModels
+          CreateViewModel = d.CreateViewModel
+          CreateCollection = d.CreateCollection >> CollectionTarget.mapVm fOut
+          UpdateViewModel = d.UpdateViewModel
+          ToMsg = d.ToMsg
+        }
+      | SubModelSeqKeyedData d -> SubModelSeqKeyedData {
+          GetSubModels = d.GetSubModels
+          CreateViewModel = d.CreateViewModel
+          CreateCollection = d.CreateCollection >> CollectionTarget.mapVm fOut
+          UpdateViewModel = d.UpdateViewModel
+          ToMsg = d.ToMsg
+          VmToId = d.VmToId
+          BmToId = d.BmToId
+        }
+      | SubModelSelectedItemData d -> SubModelSelectedItemData {
+          Get = d.Get
+          Set = d.Set
+          SubModelSeqBindingName = d.SubModelSeqBindingName
+        }
+    let rec recursiveCase<'model, 'msg, 'a, 'b> (fOut: 'a -> 'b) (fIn: 'b -> 'a) : BindingData<'model, 'msg, 'a> -> BindingData<'model, 'msg, 'b> =
+      function
+      | BaseBindingData d -> d |> baseCase fOut fIn |> BaseBindingData
+      | CachingData d -> d |> recursiveCase<'model, 'msg, 'a, 'b> fOut fIn |> CachingData
+      | ValidationData d -> ValidationData {
+          BindingData = recursiveCase<'model, 'msg, 'a, 'b> fOut fIn d.BindingData
+          Validate = d.Validate
+        }
+      | LazyData d -> LazyData {
+          Get = d.Get
+          Set = d.Set
+          BindingData = recursiveCase<obj, obj, 'a, 'b> fOut fIn d.BindingData
+          Equals = d.Equals
+        }
+      | AlterMsgStreamData d -> AlterMsgStreamData {
+          BindingData = recursiveCase<obj, obj, 'a, 'b> fOut fIn d.BindingData
+          AlterMsgStream = d.AlterMsgStream
+          Get = d.Get
+          Set = d.Set
+        }
+    
+  let boxVm b = VmMapperHelpers.recursiveCase box unbox b
```

Comment:
It works exactly the same as the `boxMinorTypes` function. By that argument, I'd have to rename that function as well.

The reason I chose `boxVm` is because its signature is `BindingData<'model,'msg,'a> -> 'BindingData<'model,'msg,obj>`, so it's boxing the view model type parameter. As another concrete example, `SubModelData.boxMinorTypes` has the signature `SubModelData<'model,'msg,'bindingModel,'bindingMsg,'vm> -> SubModelData<'model,'msg,obj,obj,'vm>` and still is accurately named, despite the fact that it actually unboxes more often than it boxes in its implementation.

They still result in strong types going in and boxed types going out, so it's box.

Maybe you want it renamed to `boxVmType` to keep it more congruent with `boxMinorTypes`?

Reply:
I agree with the `box` part of the name.  I am disagreeing with the `Vm` part of the name.  Sorry that that wasn't clear.

> The reason I chose `boxVm` is because its signature is `BindingData<'model,'msg,'a> -> 'BindingData<'model,'msg,obj>`, so it's boxing the view model type parameter.

There is no view model in the `OneWay`, `OneWayToSource`, and `TwoWay` cases, so the type parameter in those cases is not a view model type parameter.

Reply:
`'a` is the type of the view model property, not the type of a submodel (it *may* be that in the case of `SubModel` bindings, and in the case of `SubModelSeq...` bindings it's `ObservableCollection<'viewModelItem>`). So it is in fact used in `OneWay`, `OneWayToSource`, and `TwoWay` cases.

Reply:
I don't understand.  Maybe we should meet to discuss this.

Reply:
Sorry I worded that horribly. `'a` is the type of the property on the view model. It's the type that `OneWay` returns to the view model and `OneWayToSource` accepts on a set from the view model. In addition, it's also the concrete type that `SubModel` returns, which happens to be another view model.

`boxVm` refers to the fact that it's boxing just the property on the view model, not anything else. But it's referring to `'a` for OneWay, TwoWay, and OneWayToSource, `'vm` for SubModel and SubModelWin, and `'vmCollection` for OneWaySeq and SubModelSeq.

Reply:
@TysonMN I renamed everything to use "T" since that's the marker going forward. It's now `'t` in all the places where we *know* that it's the type of the property. Lazy, Cached BindingData, etc use `'t` because that's all they know about it. BaseBindingData *chooses* which is the `'t` type parameter. `'t` is then used everywhere else (BindingVmHelpers' Get, Set, Initialize, etc).

But for OneWay, TwoWay, OneWayToSource it's still `'a`, SubModel and SubModelWin it's `'vm`, OneWaySeq is `'a` *and* `'aCollection`, and SubModelSeqKeyed/Unkeyed are `'vm` *and* `'vmCollection`. Because at those levels, we don't actually "know" which one `BaseBindingData` is going to "choose" to be the `'t`

So now the new name of the function is `boxT` which is very nice considering all of the static stuff is going to be `BindingT`.

Additionally I renamed the types for CollectionTarget to reflect that they are `'a` and `'aCollection` since we are saying nothing specific about what's inside the collection.

Reply:
Resolving since I decided on `'t` and `T` when referencing to the overall type of view model property.